### PR TITLE
Add time based UUID generation component

### PR DESCRIPTION
**Pull request by denuwanthi**
## Purpose
Add component to generate time based UUID.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.carbon.uuid.generator.*,
+                        </Export-Package>
+                        <Import-Package>
+                            !org.wso2.carbon.uuid.generator.*,
+                            *;resolution:=optional
```
Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml:48]: Let's not have resolution optional. Import what's necessary

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            <artifactId>org.apache.felix.scr.ds-annotations</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.core</artifactId>
+        </dependency>
+    </dependencies>
+
+</project>
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml:68]: Add a new line here

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.felix</groupId>
+            <artifactId>org.apache.felix.scr.ds-annotations</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.core</artifactId>
```
Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml:64]: so does this means commons have a dependency for kernel

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+package org.wso2.carbon.uuid.generator;
+
+import org.wso2.carbon.uuid.generator.util.UUIDGeneratorConstants;
+import org.wso2.carbon.uuid.generator.util.Util;
+
+import java.security.SecureRandom;
+import java.util.Random;
+import java.util.UUID;
+
+public class TimeBasedUUIDGenerator extends UUIDGenerator{
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:27]: Check formatting..

Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:27]: class comment

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    long _address;
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:29]: Shall we refactor the variable name?

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        //set clock_seq_low.
+        timeBaseduuidByteAray[UUIDGeneratorConstants.CLOCK_SEQUENCE_LOW_BYTE_OFFSET_IN_UUID] = (byte) clockSequence;
+        long clockSeqInLong = Util.createClockSeqInLong(timeBaseduuidByteAray, 8);
+        uuidSecondHalf = Util.createUUIDSecondHalfInLong(clockSeqInLong);
+        uuidTimeStamp = timeStamp;
+    }
+
+    @Override
+    public UUID generate()
+    {
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:53]: Check formatting.

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            random.nextBytes(nodeIdByteArray);
+        }
+        nodeIdByteArray[0] |= (byte) 0x01;
+        createBinaryAddress(nodeIdByteArray);
+        return nodeIdByteArray;
+    }
+
+    private void createBinaryAddress(byte[] nodeIdByteArray){
+        if (nodeIdByteArray.length != 6) {
+            throw new NumberFormatException("Ethernet address has to consist of 6 bytes");
```
Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:85]: What's the relation with the ethernet address

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        midhi |= 0x1000; // type 1
+        long midhiL = (long) midhi;
+        midhiL = ((midhiL << 32) >>> 32); // to get rid of sign extension
+        // and reconstruct
+        long l1 = (((long) clockLo) << 32) | midhiL;
+        // last detail: must force 2 MSB to be '10'
+        return new UUID(l1, uuidSecondHalf);
+    }
+
+    private byte[] constructNodeId(){
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:71]: Check formatting..

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        byte[] nodeIdByteArray = new byte[6];
+        synchronized (random) {
+            random.nextBytes(nodeIdByteArray);
+        }
+        nodeIdByteArray[0] |= (byte) 0x01;
+        createBinaryAddress(nodeIdByteArray);
+        return nodeIdByteArray;
+    }
+
+    private void createBinaryAddress(byte[] nodeIdByteArray){
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:83]: Check formatting..

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        long l = nodeIdByteArray[0] & 0xFF;
+        for (int i = 1; i < 6; ++i) {
+            l = (l << 8) | (nodeIdByteArray[i] & 0xFF);
+        }
+        _address = l;
+    }
+
+    private void insertToByteArray(byte[] array, int pos)
+    {
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:95]: Check formatting..

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        if (pos < 0 || (pos + 6) > array.length) {
+            throw new IllegalArgumentException("Illegal offset ("+pos+"), need room for 6 bytes");
```
Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:97]: format

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
+/* JUG Java Uuid Generator
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java]: Add the WSO2 license here

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+ * Licensed under the License specified in the file LICENSE which is
+ * included with the source code.
+ * You may not use this file except in compliance with the License.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Comment (malithie) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java:17]: License ? we might have to recheck this it seems

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDGenerator.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.wso2.carbon.uuid.generator;
+
+import java.util.UUID;
+
+public abstract class UUIDGenerator {
+
+    public abstract UUID generate();
+
+
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDGenerator.java:26]: Remove unnecessary new lines.

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDTimeStamp.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.uuid.generator;
+
+import java.io.*;
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDTimeStamp.java:20]: Let's remove the wild card imports.

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/util/Util.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+ * under the License.
+ */
+package org.wso2.carbon.uuid.generator.util;
+
+public class Util {
+
+    public final static long createClockSeqInLong(byte[] buffer, int offset)
+    {
+        long hi = ((long) _gatherInt(buffer, offset)) << 32;
+        //long lo = ((long) _gatherInt(buffer, offset+4)) & MASK_LOW_INT;
```
Comment (ashensw) [inline components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/util/Util.java:25]: Shall we remove commented lines?

File: components/uuid-generator/pom.xml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    <artifactId>uuid-generator</artifactId>
+    <packaging>pom</packaging>
+    <name>WSO2 Carbon - UUID Generator Aggregator Module</name>
+    <url>http://wso2.org</url>
+
+    <modules>
+        <module>org.wso2.carbon.uuid.generator</module>
+    </modules>
+
+</project>
```
Comment (ashensw) [inline components/uuid-generator/pom.xml:37]: Add a new line.

File: features/uuid-generator/org.wso2.carbon.uuid.generator.feature/pom.xml [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+                                </includedFeatureDef>
+                            </includedFeatures>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>
```
Comment (ashensw) [inline features/uuid-generator/org.wso2.carbon.uuid.generator.feature/pom.xml:70]: Add a new line.

File: features/uuid-generator/org.wso2.carbon.uuid.generator.server.feature/pom.xml [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+                            </bundles>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+
+        </plugins>
+    </build>
+
+</project>
```
Comment (ashensw) [inline features/uuid-generator/org.wso2.carbon.uuid.generator.server.feature/pom.xml:76]: Add a new line.

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 7 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java:85]: This should be node id address.
According to RFC this is the description for Node id

 For UUID version 1, the node field consists of an IEEE 802 MAC
   address, usually the host address.  For systems with multiple IEEE
   802 addresses, any available one can be used.  The lowest addressed
   octet (octet number 10) contains the global/local bit and the
   unicast/multicast bit, and is the first octet of the address
   transmitted on an 802.3 LAN.

   For systems with no IEEE address, a randomly or pseudo-randomly
   generated value may be used; see Section 4.5.  The multicast bit must
   be set in such addresses, in order that they will never conflict with
   addresses obtained from network cards.

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml (hunk 1 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml]: Fixed with 05f8b087ce8b2b579ff28c668dafa4ad52ee0a8f

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml (hunk 3 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml]: No we don't need this dependecy.
Fixed with 05f8b087ce8b2b579ff28c668dafa4ad52ee0a8f

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 4 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 7 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 11 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml (hunk 2 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/pom.xml]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 4 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 5 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 6 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 8 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 9 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java (hunk 10 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java (hunk 12 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimestampSynchronizer.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDGenerator.java (hunk 14 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDGenerator.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDTimeStamp.java (hunk 15 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/UUIDTimeStamp.java]: Fixed with 05f8b08

File: components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/util/Util.java (hunk 16 above)
Comment (denuwanthi) [reply components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/util/Util.java]: Fixed with 05f8b08

File: components/uuid-generator/pom.xml (hunk 17 above)
Comment (denuwanthi) [reply components/uuid-generator/pom.xml]: Fixed with 05f8b08

File: features/uuid-generator/org.wso2.carbon.uuid.generator.feature/pom.xml (hunk 18 above)
Comment (denuwanthi) [reply features/uuid-generator/org.wso2.carbon.uuid.generator.feature/pom.xml]: Fixed with 05f8b08

File: features/uuid-generator/org.wso2.carbon.uuid.generator.server.feature/pom.xml (hunk 19 above)
Comment (denuwanthi) [reply features/uuid-generator/org.wso2.carbon.uuid.generator.server.feature/pom.xml]: Fixed with 05f8b08

**Pull request merged**